streamr-dev/streamr-platform
# [PLATFORM-1145] Canvas message log sidebar

Following a discussion with Matt, this sidebar now only contains validation messages rather than validation messages and a log of the notifications sent, at least for this first version. It wasn't immediately clear how the validation messages should/could work within an event log.

 Main motivation for this feature is to start being able to help users with common issues e.g. selecting a stream with no fields configured. Realtime log much lower value than being able to provide these messages.

Needs some additional design feedback e.g. for the message grouping headers, which I will collect from Matt if/when the PR deploys.

Added frontend "validation" for:

* Stream module with no stream selected.
* Stream module with a stream selected but that stream has no fields configured.
* Also grabbed module errors from `canvas.moduleErrors`. Afaik only the Java module's compile errors appear here at the moment.

Going forward we should probably produce warnings/errors for at least:

* Solidity Module with no account selected
* Canvas Module with no canvas selected
* And the solidity module should probably also use the same `canvas.moduleErrors` like the Java module.

So while we can add additional validation on the frontend, it's possible we should actually look at updating e&e so it produces the above messages into `canvas.moduleErrors`, and perhaps also change the name from `moduleErrors` to `moduleMessages` or even just `messages`. 

Perhaps also need backend to start adding an additional flag for inputs/params that require a value.

#### Also

* Dedupes module names e.g. "Stream" -> "Stream 02" -> "Stream 03". No deduping if you edit module name manually, only on module creation. This is so modules can be differentiated in sidebar.
* Adds "Copy" to name of copied modules
* Uses "smart" offset finding when pasting. Before it just added 32px offset from original position, so didn't stack.
* Immediately "applies" when opening the Java module, this allows annotations to appear.
* Might have fixed at least one instance of that weird scrolled canvas issue @fonty1 mentioned.

#### To Test

1. Create Stream module
2. Note badge on stream module and on console icon in toolbar
3. Clicking either badge or console icon in toolbar opens console sidebar
4. Console sidebar should say "no stream selected"
5. Select any Stream
6. Error in sidebar should go away if selected stream has fields
7. Create a stream with no fields
8. Select that stream
9. Console sidebar should complain that stream has no fields, and provide a link to edit stream.
10. Add Java module
11. Edit code so there's multiple errors. Errors should appear in console sidebar.
12. Selecting a module with the console sidebar open should scroll to that module's errors in the console sidebar.
13. Selecting an error in the sidebar should select & pan to the associated module. Note that currently "pan to selected module" doesn't take sidebar position into account, so panned to module may be behind sidebar.
14. Copy/pasting a module will dedupe the name, and find appropriate offset position. This name is used in the sidebar to disambiguate modules.

Also note if you're playing around with this:

* If there's errors in the Java module, Stream modules won't update their ports from the backend until the Java module errors are fixed.
* You can't copy/paste a Java module that has errors, gives a null pointer exception on server.

----

![image](https://user-images.githubusercontent.com/43438/70128548-001e4a00-16b8-11ea-81fb-d63e9eff0cf1.png)

![image](https://user-images.githubusercontent.com/43438/70128722-55f2f200-16b8-11ea-8dff-4fd1fcaf294d.png)
